addon/components/frost-popover.js

```diff
@@ -1,96 +1,328 @@
 import Ember from 'ember'
 import layout from '../templates/components/frost-popover'
 import $ from 'jquery'
-export default Ember.Component.extend({
+import PropTypeMixin, {PropTypes} from 'ember-prop-types'
+import {checkLeft, checkRight, checkTop, checkBottom} from './util'
+
+const arrowMargin = 5
+const maxPlacementRetries = 5
+
+export default Ember.Component.extend(PropTypeMixin, {
   layout,
-  visible: false,
-  event: 'click',
-  closest: false,
-  offset: 10,
-  position: 'bottom',
-  index: 0,
-  excludePadding: false,
-  classNameBindings: ['position', 'visible:visible:invisible'],
+  classNameBindings: ['visible:visible:invisible', 'autoPosition'],
   classNames: ['tooltip-frost-popover'],
-  didRender () {
+  propTypes: {
+    closest: PropTypes.bool,
+    event: PropTypes.string,
+    excludePadding: PropTypes.bool,
+    index: PropTypes.number,
+    offset: PropTypes.number,
+    position: PropTypes.string,
+    resize: PropTypes.bool,
+    viewport: PropTypes.oneOfType([
+      PropTypes.string,
+      PropTypes.object,
+      PropTypes.func
+    ])
+  },
+
+  getDefaultProps () {
+    return {
+      closest: false,
+      event: 'click',
+      excludePadding: false,
+      index: 0,
+      offset: 10,
+      position: 'bottom',
+      resize: true,
+      viewport: 'body',
+      visible: false,
+      autoPosition: 'bottom'
+    }
+  },
+
+  didInsertElement () {
     Ember.run.next(() => {
-      const context = this
-      if (this.get('closest')) {
-        this.$().closest(this.get('target')).on(this.get('event'), function () {
-          context.send('togglePopover')
-        })
+      const target = this.getTarget()
+      const event = this.get('event')
+
+      this._eventHandler = (event) => {
+        const popover = this.get('element')
+        if ($(event.target).closest(popover).length === 0) {
+          this.send('togglePopover')
+        }
+      }
+
+      $(target).on(event, this._eventHandler)
```

Comment:
I understand why we did this originally, but I think at this point it would probably make sense to bind to a shared state object - so the target would fire an event to mutate a state, then the popover would bind to that state and display.  I know it's a bit more boilerplate, but it allows us to manage the state from a central location.